### PR TITLE
net: dns_sd, mdns: support service type enumeration

### LINKED ISSUE #38673: DNS-SD library does not support `_services._dns-sd._udp.local` meta-query for service enumeration
**Is your enhancement proposal related to a problem? Please describe.**
When building and running the mdns_responder sample, the device won't show up in avahi-browse or zeroconfServiceBrowser automatically. The device will only reply when you specifically query the following: `avahi-browse -t -r _zephyr._tcp`.

It seems like `_services._dns-sd._udp.<domain>` meta-query handling is not implemented.

When I open up zeroconfServiceBrowser or avahi-browse, the following is printed out at regular intervals: 
`net_dns_sd.dns_sd_extract_service_proto_domain: (rx_q[0]): could not get proto for '▒▒...'`

**Describe the solution you'd like**
The device should respond properly to the meta-query `_services._dns-sd._udp.<domain>` so that my device is displayed automatically in avahi-browse without having to specifically query `avahi-browse -t -r _zephyr._tcp`

### PR DESCRIPTION
**Pull request by cfriedt**
Support DNS-SD Service Type Enumeration in the dns library and mdns_responder sample application.
    
For more information, please see Section 9, "Service Type Enumeration" in RFC 6763.
    
https://datatracker.ietf.org/doc/html/rfc6763#section-9

Fixes #38673

### CODE REVIEW

File: subsys/net/lib/dns/mdns_responder.c
Comment (cfriedt) [inline subsys/net/lib/dns/mdns_responder.c:340]: `dns_sd_extract_service_proto_domain()` will need to be modified to also *optionally* extract instance.

* pass in an `instance_buf`
* if non-`NULL`, and size is OK, then extract instance
* in the end, will need to fill 3, maybe 4 buffers, so a bit of a rework will be required.

File: subsys/net/lib/dns/dns_sd.c
Comment (jukkar) [inline subsys/net/lib/dns/dns_sd.c:28]: Shouldn't we set this to 0 here?
Also wondering whether the `dns_sd_empty_txt` is cleared somewhere in the code.

File: include/net/dns_sd.h
Comment (jukkar) [inline include/net/dns_sd.h:218]: Perhaps this has been discussed earlier already, but just wondering why do we need to expose these extern variables in a public header file. Better would be to place any internal stuff to a private header file that could be placed to dns directory directly.

File: subsys/net/lib/dns/dns_sd.c
Comment (cfriedt) [reply subsys/net/lib/dns/dns_sd.c:28]: It's set to zero as part of bss (uninitialized global variables)

File: include/net/dns_sd.h
Comment (cfriedt) [reply include/net/dns_sd.h:218]: All dns_sd_rec are placed in an iterable struct section which requires a `uint16_t*` to facilitate ephemeral ports. It cannot be NULL atm. Rather than having several global `uint16_t` floating around solely for this purpose, I thought it best to make 1.

File: subsys/net/lib/dns/dns_sd.c
Comment (jukkar) [reply subsys/net/lib/dns/dns_sd.c:28]: Hmm, I seem to have missed the memo that bss is initialized to zero in zephyr (as the name says it contains uninitialized vars thus was expecting `=0` here).

File: include/net/dns_sd.h
Comment (jukkar) [reply include/net/dns_sd.h:218]: > Rather than having several global `uint16_t` floating around solely for this purpose, I thought it best to make 1.

Sure, but my question was that do we need to have these global symbols exposed in public header file, could these be placed in dns-sd private header file instead.

Comment (cfriedt) [reply include/net/dns_sd.h:218]: > > Rather than having several global `uint16_t` floating around solely for this purpose, I thought it best to make 1.
> 
> Sure, but my question was that do we need to have these global symbols exposed in public header file, could these be placed in dns-sd private header file instead.

They need to be in a public header so that API consumers can use `DNS_SD_REGISTER_SERVICE()`, `DNS_SD_REGISTER_TCP_SERVICE()`, `DNS_SD_REGISTER_UDP_SERVICE()`, etc.

Comment (jukkar) [reply include/net/dns_sd.h:218]: Ok, no worries, sorry for the noise.

File: subsys/net/lib/dns/dns_sd.c
Comment (cfriedt) [reply subsys/net/lib/dns/dns_sd.c:28]: Yea, bss is set to zero as part of the C standard (C89). Zephyr follows suit ;-)

**Comment by cfriedt** [comment]
The compliance failure is a false positive (bug in `./scripts/checkpatch.pl`. @galak, you may need to merge this one manually after it's approved.

File: subsys/net/lib/dns/mdns_responder.c
Comment (cfriedt) [reply subsys/net/lib/dns/mdns_responder.c]: Filed issue #38846

File: tests/net/lib/dns_sd/src/main.c
Comment (cfriedt) [inline tests/net/lib/dns_sd/src/main.c:714]: These numbers come from the query for `_services._dns-sd._udp.local`
Screen Shot 2021-09-23 at 7 15 02 AM

**Comment by cfriedt** [comment]
Works now with `avahi-browse -r -t -a
Screen Shot 2021-09-26 at 6 58 52 AM
`

File: subsys/net/lib/dns/mdns_responder.c
Comment (cfriedt) [reply subsys/net/lib/dns/mdns_responder.c:369]: should be fixed in latest revision. Will test with `avahi-discover` shortly.

**Comment by cfriedt** [comment]
So service type enumeration is working now, in a way that wireshark agrees with. However, I do need to do a `avahi-browse -r -t -a` in order for avahi to send out the enumeration query. `avahi-discover` *does* send out the query but it seems that 'non-standard' services are not listed. I was able to test with `BOARD=stm32f746g_disco` and the macOS application `discovery` and `_zephyr._tcp.local` is indeed automatically discovered with this fix.
Screen Shot 2021-09-27 at 12 53 36 PM
Screen Shot 2021-09-27 at 12 14 06 PM
Screen Shot 2021-09-27 at 12 13 03 PM

**Comment by cfriedt** [comment]
Also, note, that there is an enhancement ticket to provide an overall improvement of DNS-SD in v3.0.0, with emphasis on optional features like this (service type enumeration), domain enumeration, as well as complete name, subtype, and non-local domain support.

For more information, see #38846

**Comment by cfriedt** [comment]
@rlubos , @jukkar - please have a look when you have a few minutes.

File: subsys/net/lib/dns/dns_sd.c
Comment (jukkar) [inline subsys/net/lib/dns/dns_sd.c:832]: I think we should make sure that `sizeof(*rsp) <= buf_size` before calling `memset`, so something like
```
if (sizeof(*rsp) > (size_t)buf_size) {
     return -EINVAL;
}
```

File: subsys/net/lib/dns/dns_sd.c
Comment (jukkar) [inline subsys/net/lib/dns/dns_sd.c:841]: The `ntohs()` looks unnecessary as the value is 0, so you can replace it by literal `0` value

File: subsys/net/lib/dns/dns_sd.c
Comment (jukkar) [inline subsys/net/lib/dns/dns_sd.c:843]: This value looks slightly weird, wouldn't `EINVAL` be more appropriate here or if we want to have network specific value, then perhaps `EADDRNOTAVAIL` could be used here

File: subsys/net/lib/dns/dns_sd.c
Comment (jukkar) [inline subsys/net/lib/dns/dns_sd.c:857]: Perhaps `EADDRINUSE` here

File: subsys/net/lib/dns/dns_sd.c
Comment (jukkar) [inline subsys/net/lib/dns/dns_sd.c:1000]: We are checking int and size_t here, this could cause issues if value of N is very large. Better to add a cast `(int)N` here to in order to avoid some Coverity issue later.

File: subsys/net/lib/dns/dns_sd.c
Comment (jukkar) [inline subsys/net/lib/dns/dns_sd.c:1042]: same here, please use `(int)N`

File: subsys/net/lib/dns/dns_sd.c
Comment (jukkar) [inline subsys/net/lib/dns/dns_sd.c:1056]: ditto

File: subsys/net/lib/dns/dns_sd.c
Comment (jukkar) [inline subsys/net/lib/dns/dns_sd.c:1073]: ditto

File: tests/net/lib/dns_sd/src/main.c
Comment (jukkar) [inline tests/net/lib/dns_sd/src/main.c:522]: Adding `{}` is not needed as the array is already cleared to 0 (because it is static)

File: subsys/net/lib/dns/dns_sd.c
Comment (cfriedt) [reply subsys/net/lib/dns/dns_sd.c:841]: This matches what is done for PTR handling, so if it were changed here it would need to be changed there as well. I would prefer to address this in the larger rework task for v3

File: subsys/net/lib/dns/dns_sd.c
Comment (cfriedt) [reply subsys/net/lib/dns/dns_sd.c:843]: This matches what is done for PTR handling, so if it were changed here it would need to be changed there as well. I would prefer to address this in the larger rework task for v3

File: subsys/net/lib/dns/dns_sd.c
Comment (cfriedt) [reply subsys/net/lib/dns/dns_sd.c:857]: This matches what is done for PTR handling, so if it were changed here it would need to be changed there as well. I would prefer to address this in the larger rework task for v3

File: subsys/net/lib/dns/dns_sd.c
Comment (rlubos) [inline subsys/net/lib/dns/dns_sd.c:840]: No need to break the string I guess? It should still fit within the limit.

File: subsys/net/lib/dns/dns_sd.c
Comment (rlubos) [inline subsys/net/lib/dns/dns_sd.c:1080]: What if we receive a query with an unsupported "label" count? Shouldn't we return an error in such case (record will not be set then)? Even if that's not supported by the standard, we shall still be ready fol malformed packets.

**Review by rlubos — COMMENTED** [review]
Not much to add, looks quite ok. I confirm that Zephyr will now reply for Avahii enumarate request (but as already stated for whatever reason Zephyr service does not get listed,but it could be Avahii thing, the response is on the wire).

File: subsys/net/lib/dns/dns_sd.c
Comment (cfriedt) [reply subsys/net/lib/dns/dns_sd.c:1080]: gah - sorry, I originally had 
```
default:
case DNS_SD_MAX_LABELS:
```
otherwise, for `qlabels < DNS_SD_MAX_LABELS`, it's an error

Comment (cfriedt) [reply subsys/net/lib/dns/dns_sd.c:1080]: Currently min is 3 and max is 4

**Review by rlubos — APPROVED** [review]
Thanks

**Comment by cfriedt** [comment]
@galak / @nashif - either of you want to ack? Compliance failure is a false positive, so this needs to bee merged manually.

**Review by jukkar — APPROVED** [review]

**Comment by cfriedt** [comment]
Added workaround for bug in compliance checks.

**Comment by AllemanMa** [comment]
When I performed ` dns-sd -B _zephyr._tcp` on my pc to query my device running Zephyr v2.6.0, I was able to get a response  like below containing all the necessary information.

image

If I try the same query with Zephyr v2.7.0-rc4 with the fix for the service type enumeration, my device seems to not respond the second packet anymore....

**Pull request merged**
